File: Problems & Solutions/L/Largest Sum Cycle/02 February 2023/Largest Sum Cycle.py
import sys
sys.setrecursionlimit(10**6)

class Solution():
    def dfs(self, node,p=-1):
        self.vis[node]=1
        self.par[node]=p
        self.tmp.append(node)
        for i in self.v[node]:
            if(self.vis[i]==0):
                z=self.dfs(i,node)
                if(z!=-1):
                    return z
                
            elif(self.vis[i]==1):
                sum=i
                while(node!=i):
                    sum+=node
                    node=self.par[node]
                
                if(node==i):
                    return sum
                return -1
            
        return -1

    def largestSumCycle(self, N, Edge):
        ans=-1
        self.vis=[0]*N
        self.v=[[] for i in range(N)]
        self.par = [0]*N
        self.tmp = []

        for i in range(N):
            if(Edge[i]!=-1):
                self.v[i].append(Edge[i])

        for i in range(N):
            if(~self.vis[i]):
                ans=max(ans,self.dfs(i))
                for j in self.tmp:
                    self.vis[j]=2
                
                self.tmp.clear()
        return ans

#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == "__main__":
    for _ in range(int(input())):
        N = int(input())
        Edge=[int(i) for i in input().split()]
        print(Solution().largestSumCycle(N, Edge))
# } Driver Code Ends

File: Problems & Solutions/L/Largest Sum Cycle/02 February 2023/Largest Sum Cycle.cpp
//{ Driver Code Starts
//Initial Template for C++
#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends
//User function Template for C++
class Solution
{
  public:
  
  long long largestcycle=-1;
  
  void DFS(int node,vector<int>& Edge,bool visited[],bool pathvisited[])
  {
      visited[node]=true;
      pathvisited[node]=true;
      
      if(Edge[node]!=-1)
      {
          int adj=Edge[node];
          if(!visited[adj])
          {
              DFS(adj,Edge,visited,pathvisited);
          }
          else if(pathvisited[adj])
          {
              int curr=adj;
              long long sum=0;
              do
              {
                  sum+=curr;
                  curr=Edge[curr];
              }while(curr!=adj);
              largestcycle=max(largestcycle,sum);
          }
      }
      
      pathvisited[node]=false;
  }
  
  long long largestSumCycle(int N, vector<int> Edge)
  {
    // code here
    bool visited[N]={0};
      bool pathvisited[N]={0};
      for(int i=0;i<N;i++)
      {
          if(!visited[i])
          {
              DFS(i,Edge,visited,pathvisited);
          }
      }
      return largestcycle;
  }
};

//{ Driver Code Starts.
signed main(){
   int tc;
   cin >> tc;
   while(tc--){
      int N;
      cin >> N;
      vector<int> Edge(N);
      for(int i=0;i<N;i++){
        cin>>Edge[i];
      }
      Solution obj;
      long long ans=obj.largestSumCycle(N, Edge);
      cout<<ans<<endl;
   }
   return 0;
}
// } Driver Code Ends
